### PR TITLE
[Bug] Fix root cause that caused StatsGranularTabsTest to fail.

### LINKED ISSUE #18065: Very first Stats card is not always loaded on screen
### Expected behavior

All stats cards that can be shown on screen should be shown after opening the screen, no PTR should be required.

### Actual behavior

1. In some cases, for JP variant (I could not reproduce this on WP), the very first Stats card is loaded only after PTR:

https://user-images.githubusercontent.com/73365754/223773395-9ed971b6-c145-4a6b-a39a-2796d9c81332.mov

2. In other cases ([example](https://buildkite.com/automattic/wordpress-android/builds/10502#01876bb0-cb7e-4fa4-bd0d-d9d8fbbea206)) the stats screen does not load at all.

This causes our `e2eAllDayStatsLoad` UI test to fail for JP.

I could not reproduce this on a real device yet.

### Steps to reproduce the behavior

1. Launch JP
2. Log in and open `Stats`
3. Open `Days` tab ~> in some cases, the first card with bar chart is loaded only after PTR.

##### Tested on `Pixel 2 API 31 Emulator`, `JPAndroid 21.8-rc-2`

### PR DESCRIPTION
**Pull request by notandyvee**
Fixes #18065

There was a failing UI test that appeared flaky, `StatsGranularTabsTest`. Retrying appears to fix it. Every once in a while I also notice stats cards don't load properly. It's super hard to reproduce, even in UI tests. 

At first I figured out part of the issue. I noticed each failing stats card would have an empty `BlockListItem` list. Each card can have multiple `BlockListItem`s. When the `StatsBlock` has no `BlockListItem`, it can't do anything, so presents itself as a card with nothing in it. This is unfortunate, but not really an issue. The real issue is why a `StatsBlock` can even have an empty list.

After further testing I learned that what leads to this issue is actually a race condition. The UI is dependent on setting the `domainModel` in `BaseStatsUseCase`. In most cases, the domain model is present quickly when we load the UI. But during UI tests situations arise where the `domainModel` takes a bit longer to load, leading to a race condition. As far as I can tell, it really only happens (or noticeable) when you first login or switch to a blog for the first time. The change I made in `BaseStatsUseCase` best illustrates why I think this fixes the issue.

When a user first loads stats, there is no cache. So it must fetch from the network. When it does that, the `domainModel` is set to the cache, which technically doesn't exist yet. This happens in `evaluateState` when a network call is made.

```
domainModel = updatedCachedData
```

It's not entirely clear to me why this is being done. In `evaluateState` if we got state back, presumably we should use that as it would be the most up-to-date. So I just tried it.

```
domainModel = state.model
```

I re-ran the previously failing UI test dozens of times. Couldn't reproduce it. I also 🚬 tested stats and don't notice any differences.

-----

## To Test:

- [ ] `StatsGranularTabTest` must pass 100% of the time. If on first pass it fails, this PR is pointless 😅 .
- [ ] 🚬 test granular stats. Make sure you load all cards and that everything looks as it did before.

-----

## Regression Notes

1. Potential unintended areas of impact

    Stats cards loading.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    Manual test and UI test.

3. What automated tests I added (or what prevented me from doing so)

    n/a

-----

## PR Submission Checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt [hunk 1]
```diff
@@ -94,7 +94,7 @@ abstract class BaseStatsUseCase<DOMAIN_MODEL, UI_STATE>(
                 if (!state.cached) {
                     val updatedCachedData = loadCachedData()
                     if (domainModel != updatedCachedData) {
-                        domainModel = updatedCachedData
+                        domainModel = state.model
```
Comment (ravishanker) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt]: Does it mean the whole if{} block is redundant here?!

Comment (notandyvee) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt]: At first glance, the answer is that its not redundant. Not because of my change, but because of the `updateState()`. If the state values are the same, I imagine we don't want to fire a call to `updateState`. But this is where your input will matter as there may be details about fluxc I don't fully understand. Presumably, if we are using a data class, which I believe we do for those models, data from the cache can equal data from the web. Basing this solely off the [docs](https://kotlinlang.org/docs/data-classes.html). I could be wrong. Your thoughts 😄 ?

Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt]: FluxC is also long before my time!

Wdyt? @irfano

Comment (notandyvee) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt]: You know what. I also just realized that I see why you are saying that. `domainModel != updatedCachedData` is confusing. It should be `domainModel != state`. Does that make more sense @ravishanker ?

Comment (irfano) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt]: > You know what. I also just realized that I see why you are saying that. domainModel != updatedCachedData is confusing. It should be domainModel != state. Does that make more sense @ravishanker ?

Yes! We can remove `val updatedCachedData = loadCachedData()` line and check `domainModel != state`

**Review by irfano — APPROVED** [review]
> When a user first loads stats, there is no cache. So it must fetch from the network. When it does that, the domainModel is set to the cache, which technically doesn't exist yet. 

This is not entirely accurate. FluxC [writes to the cache](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/fluxc/src/main/java/org/wordpress/android/fluxc/store/stats/time/VisitsAndViewsStore.kt#L68) before sending responses to the WordPress-Android. So, before `val updatedCachedData = loadCachedData()` line, it is already fetched and the cache has existed. But I do agree with your final suggestion.

> It's not entirely clear to me why this is being done. In evaluateState if we got state back, presumably we should use that as it would be the most up-to-date. So I just tried it.

I agree!

I tested `trunk` and this PR multiple times and have never encountered the issue on this PR while I could reproduce it on `trunk`. Additionally, I compared the content of all the cards between PR and the `trunk` version, and they were identical.
Andy, you can address Ravi's comment, and then I believe we're good to merge.  🟢

**Comment by notandyvee** [comment]
Thanks for adding the missing context @irfano . Very useful information. And actually makes me wonder why we don't just listen to the database cache updates directly and not deal with the network at all 😅 . But that's out of scope for this PR. I went ahead and made that change. Ended up removing the `if` like Ravi said anyways.

**Comment by notandyvee** [comment]
Damn. Heads up @irfano . That seemingly small change causes a few new failed tests. Some of them are expected. But others are tricky, like [this one](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/detail/PostDayViewsUseCaseTest.kt#L139). I need to slowly go through it all and ensure I am not introducing a regression 😅 .

**Review by ravishanker — APPROVED** [review]
Great work 🚀

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-Android/milestone/281">24.7</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by notandyvee** [comment]
Hey @irfano when you get a chance, no rush. Quick re-review. I made the changes we talked about. But had to update a bunch of unit tests. I may have went down a rabbit hole. But the error wasn't clear here, so I went the extra mile and updated the unit tests so they pass locally without the need to change the flavor, which is more proper anyways.

Note: Most of the files edited were test files. Small exceptions.

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/detail/PostDayViewsUseCaseTest.kt [hunk 2]
```diff
@@ -139,7 +141,6 @@ class PostDayViewsUseCaseTest : BaseUnitTest() {
     fun `manage edge condition with data available but empty list`() = test {
```
Comment (irfano) [inline WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/detail/PostDayViewsUseCaseTest.kt]: Does `manage edge condition with data available but empty list` still have a valid test name and comment?

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCaseTest.kt [hunk 3]
```diff
@@ -55,19 +55,19 @@ class BaseStatsUseCaseTest : BaseUnitTest() {
         block.fetch(false, false)
         advanceUntilIdle()
 
-        assertData(0, localData)
+        assertData(0, remoteData)
```
Comment (irfano) [inline WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCaseTest.kt:58]: We should also update the test function name to `on fetch loads data from remote when current value is null`

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCaseTest.kt [hunk 4]
```diff
@@ -55,19 +55,19 @@ class BaseStatsUseCaseTest : BaseUnitTest() {
...
     }
 
     @Test
-    fun `on fetch returns null item when DB is empty`() = test {
+    fun `on fetch uses remote data when DB is empty`() = test {
```
Comment (irfano) [inline WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCaseTest.kt]: np: Does `...returns remote data` instead of `uses remote data` sound better to you?

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCaseTest.kt [hunk 5]
```diff
@@ -79,15 +79,15 @@ class BaseStatsUseCaseTest : BaseUnitTest() {
         advanceUntilIdle()
 
         assertThat(result.size).isEqualTo(1)
-        assertData(0, localData)
+        assertData(0, remoteData)
```
Comment (irfano) [inline WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCaseTest.kt:82]: We should also update test test name.

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/OverviewUseCaseTest.kt [hunk 6]
```diff
@@ -135,7 +135,8 @@ class OverviewUseCaseTest : BaseUnitTest() {
             assertThat(this[1]).isEqualTo(barChartItem)
             assertThat(this[2]).isEqualTo(columns)
         }
-        verify(statsWidgetUpdaters, times(2)).updateViewsWidget(siteId)
+        verify(statsWidgetUpdaters, times(1)).updateViewsWidget(siteId)
+        verify(statsWidgetUpdaters, times(1)).updateWeekViewsWidget(siteId)
```
Comment (irfano) [inline WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/OverviewUseCaseTest.kt:139]: ❓ Why has this change become necessary?

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCaseTest.kt [hunk 7]
```diff
@@ -150,6 +155,7 @@ class ReferrersUseCaseTest : BaseUnitTest() {
             )
         ).thenReturn(contentDescription)
         whenever(statsUtils.toFormattedString(any<Int>(), any())).then { (it.arguments[0] as Int).toString() }
+        whenever(buildConfigWrapper.isJetpackApp).thenReturn(false)
```
Comment (irfano) [inline WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCaseTest.kt:158]: Shouldn't `buildConfigWrapper.isJetpackApp` return `true` because stats is only available to Jetpack? The same applies to all other tests where `buildConfigWrapper.isJetpackApp` is used.

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/AllTimeStatsUseCaseTest.kt [hunk 8]
```diff
@@ -153,7 +153,7 @@ class AllTimeStatsUseCaseTest : BaseUnitTest() {
             assertThat(this.endColumn.value).isEqualTo(viewsBestDayTotal.toString())
             assertThat(this.endColumn.highest).isEqualTo(bestDayTransformed)
         }
-        verify(statsWidgetUpdaters, times(2)).updateAllTimeWidget(siteId)
+        verify(statsWidgetUpdaters, times(1)).updateAllTimeWidget(siteId)
```
Comment (irfano) [inline WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/AllTimeStatsUseCaseTest.kt:156]: Do you know why this is being called once now?

File: WordPress/build.gradle [hunk 9]
```diff
@@ -478,6 +478,7 @@ dependencies {
     testImplementation "org.jetbrains.kotlin:kotlin-test-junit:$gradle.ext.kotlinVersion"
     testImplementation "org.assertj:assertj-core:$assertjVersion"
     testImplementation "org.jetbrains.kotlinx:kotlinx-coroutines-test:$kotlinxCoroutinesVersion"
+    testImplementation "org.jetbrains.kotlin:kotlin-reflect:$gradle.ext.kotlinVersion"
```
Comment (irfano) [inline WordPress/build.gradle]: Why is this necessary? Tests are still passing when I remove it.

**Review by irfano — CHANGES_REQUESTED** [review]

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCaseTest.kt (hunk 7 above)
Comment (notandyvee) [reply WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCaseTest.kt:158]: Yes. But the tests assume `isJetpackApp` being false.  Apparently the stats use cases have this if/then.

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/AllTimeStatsUseCaseTest.kt (hunk 8 above)
Comment (notandyvee) [reply WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/AllTimeStatsUseCaseTest.kt:156]: Answering only here, but the same reasoning for all the other ones.

```
                if (!state.cached) {
                    val updatedCachedData = loadCachedData()
                    if (domainModel != updatedCachedData) {
                        domainModel = updatedCachedData
                        updateState()
                    }
```

That was what was removed in `evaluateState`. We were actually calling `loadCachedData` twice, which is what makes those method calls. Once to check if we have cached data. And we also called it in `evaluateState`. Now it's only called once. Let me know if that makes sense.

File: WordPress/build.gradle (hunk 9 above)
Comment (notandyvee) [reply WordPress/build.gradle]: I was having trouble with a test, and I needed it. I think I went a different route and you are probably right. I can remove it.

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/OverviewUseCaseTest.kt (hunk 6 above)
Comment (notandyvee) [reply WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/OverviewUseCaseTest.kt:139]: Ditto from below.

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/detail/PostDayViewsUseCaseTest.kt (hunk 2 above)
Comment (notandyvee) [reply WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/detail/PostDayViewsUseCaseTest.kt]: I'll double check for you.

Comment (notandyvee) [reply WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/detail/PostDayViewsUseCaseTest.kt]: Yes. Still valid the issue is unrelated to my changes. I only change the number of times `mapper` was called. Not related to the edge case. I haven't dug deeper if this could have been related. But I think it should be out of scope.

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCaseTest.kt (hunk 5 above)
Comment (notandyvee) [reply WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCaseTest.kt:82]: Name is actually still accurate. I ran the test to confirm. The problem is actually threading (advanceUntilIdle). When you call `block.fetch`, 2-4 jobs are potentially created,  one for each call to `updateState`. But the threads/jobs created don't immediately run until we call `advanceUntilIdle`.  `BaseStatsUseCase ` would need to be re-written to more accurately reflect what would happen when we have more exact control of each job/thread. At the moment I'd need to make even more changes to improve the test. More than I think should be in scope here. Does this make sense?

**Comment by notandyvee** [comment]
Ready for re-review @irfano . Let me know if you have any additional questions. All good stuff as we should questions the changes 🙏🏻

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/detail/PostDayViewsUseCaseTest.kt (hunk 2 above)
Comment (irfano) [reply WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/detail/PostDayViewsUseCaseTest.kt]: Thank you for investigating the case Andy. I should have been clearer about what I meant. 

This **was** testing a case where:
- `whenever(store.getPostDetail(site, postId)).thenReturn(emptyModel)`: DB is empty
- ` whenever(store.fetchPostDetail(site, postId, forced)).thenReturn(OnStatsFetched(model))`: endpoint returns data for "2019-10-10"
- **expects** empty for the use cases's data. _(see test name `manage edge condition with data available but empty list`)_

The reason was that we were using DB to set the use case's data. Now, we want to use the endpoint's response instead. Since this test case should no longer expect empty data, I believe it is unnecessary and can be removed. WDYT?

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCaseTest.kt (hunk 5 above)
Comment (irfano) [reply WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCaseTest.kt:82]: Thank you for your explanation. I understand your point.
I believe the test name was already incorrect prior to this PR. It was mentioning DB and API, but asserting only `localData`. Apart from what occurs in order before `advanceUntilIdle()`, I see that this test block is testing if the result is `remoteData` after calling `block.fetch(true, false)`. So, I believe the test name should be changed to `on refresh calls loads data from API`. WDYT?

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/AllTimeStatsUseCaseTest.kt (hunk 8 above)
Comment (irfano) [reply WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/AllTimeStatsUseCaseTest.kt:156]: Yes, I got it. 👍🏻 Thank you!

**Comment by irfano** [comment]
Thank you for bearing my comments, @notandyvee. 😅 I'm curious about your opinion on my two open comments. Then we're good to merge.

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/detail/PostDayViewsUseCaseTest.kt (hunk 2 above)
Comment (notandyvee) [reply WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/detail/PostDayViewsUseCaseTest.kt]: Hmmm. You know what. I see your point. I didn't want to presume what the original issue was and that this fixes it. But based on the test, I agree. Will delete the test.

File: WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCaseTest.kt (hunk 5 above)
Comment (notandyvee) [reply WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCaseTest.kt:82]: Fair enough.

**Review by irfano — APPROVED** [review]
LGTM! 👍🏻 Thank you for your contributions to improve our tests, Andy. Great work! 🚀
